go-controller/pkg/clustermanager/clustermanager.go

```diff
@@ -0,0 +1,271 @@
+		if err := cm.secondaryNetClusterControllerManager.Start(); err != nil {
+			return err
+		}
+	}
+
+	// Start and sync the watch factory to begin listening for events
+	if err := cm.wf.Start(); err != nil {
+		return err
+	}
+
+	if err := cm.defaultNetClusterController.Run(); err != nil {
+		return err
+	}
+
+	if cm.secondaryNetClusterControllerManager != nil {
+		klog.Infof("Starting multi network attach manager")
+		return cm.secondaryNetClusterControllerManager.Run(cm.stopChan)
+	}
+
+	return nil
+}
+
+// secondaryNetworkClusterControllerManager object manages the multi net-attach-def controllers.
+// It implements networkAttachDefController.NetworkControllerManager and can be used
+// by NetAttachDefinitionController to add and delete NADs.
+type secondaryNetworkClusterControllerManager struct {
+	// net-attach-def controller handle net-attach-def and create/delete network controllers
+	nadController *nad.NetAttachDefinitionController
+	client        clientset.Interface
+	ovnClient     *util.OVNClusterManagerClientset
+	kube          kube.Interface
+	watchFactory  *factory.WatchFactory
+}
+
+func newSecondaryNetworkClusterControllerManager(ovnClient *util.OVNClusterManagerClientset,
+	wf *factory.WatchFactory, recorder record.EventRecorder) *secondaryNetworkClusterControllerManager {
+	klog.Infof("Creating new multi network cluster manager")
+	kube := &kube.Kube{
+		KClient: ovnClient.KubeClient,
+	}
+	sncm := &secondaryNetworkClusterControllerManager{
+		ovnClient:    ovnClient,
+		client:       ovnClient.KubeClient,
+		kube:         kube,
+		watchFactory: wf,
+	}
+	sncm.nadController = nad.NewNetAttachDefinitionController(
+		sncm, ovnClient.NetworkAttchDefClient, recorder)
+	return sncm
+}
+
+func (sncm *secondaryNetworkClusterControllerManager) Run(stopChan <-chan struct{}) error {
+	klog.Infof("Starting net-attach-def controller")
+	return sncm.nadController.Run(stopChan)
+}
+
+// Start starts the secondary layer3 controller, handles all events and creates all needed logical entities
+func (sncm *secondaryNetworkClusterControllerManager) Start() error {
+	klog.Infof("Start secondary network controller of network")
+	return nil
+}
+
+func (sncm *secondaryNetworkClusterControllerManager) Stop() {
+	// stops each network controller associated with net-attach-def; it is ok
+	// to call GetAllControllers here as net-attach-def controller has been stopped,
+	// and no more change of network controllers
+	klog.Infof("Stops net-attach-def controller")
+	for _, oc := range sncm.nadController.GetAllNetworkControllers() {
+		oc.Stop()
+	}
+}
+
+func (sncm *secondaryNetworkClusterControllerManager) NewNetworkController(nInfo util.NetInfo,
+	netConfInfo util.NetConfInfo) (nad.NetworkController, error) {
+	klog.Infof("New net-attach-def controller for network %s called", nInfo.GetNetworkName())
+	topoType := netConfInfo.TopologyType()
+	if topoType == ovntypes.Layer3Topology {
+		stopChan := make(chan struct{})
```

Comment:
pointless stopChan

Reply:
This stopChan is used here - https://github.com/ovn-org/ovn-kubernetes/pull/3127/files#diff-f25fbc618866d4adf33f3a32c15950be312859211e56fd7c43d13f78a594cb6eR129

A stop channel is created for every secondary network.  Is that over kill ?

Same is done here too for every secondary layer 3 network - https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/secondary_layer3_network_controller.go#L206